### PR TITLE
Update doc for API clients release policy

### PR DESCRIPTION
**Pull request by pierrejeambrun**
Following the discussion on the dev mailing list, update our documenentation (README + RELEASE PROCESS) to reflect that.

I added the API client release as a `Publish the final Apache Airflow release` step. Is this ok or do we want to be able to vote and test against API clients and therefore be included when preparing the RC ?

I also tried to include the version update of our OpenAPI spec to the release process. Let me know if this makes sense.

[Dicussion](https://lists.apache.org/thread/phbrfg2q2lnrb9d3kfz3sxv6mzvtlwvp)
[Lazy Consensus](https://lists.apache.org/list?dev@airflow.apache.org:lte=1M:lazy%20consensus)

Best,

### CODE REVIEW

**Comment by potiuk** [comment]
I think this is something that should bae tried next time when we relase airflow @ephraimbuddy @jedcunningham - I think the exact steps wil have to be run and checked when releasing Airflow 2.5.1 (then we could release API client 2.5.0)?

**Review by msumit — APPROVED** [review]

File: dev/README_RELEASE_AIRFLOW.md
Comment (eladkal) [inline dev/README_RELEASE_AIRFLOW.md:45]: I think this should be done after the Airflow release is finished?
Clients also require RC, voting.. I'm not sure if we should do all at the same time?

File: README.md
Comment (XD-DENG) [inline README.md]:
```suggestion
  As a consequence, each API client will have its own PATCH version that may or may not be in sync with the Airflow
```

nit.

**Review by potiuk — APPROVED** [review]

File: dev/README_RELEASE_AIRFLOW.md
Comment (potiuk) [reply dev/README_RELEASE_AIRFLOW.md:45]: I think in this case we should VOTE on a bundle of the Airflow + Client. There is no particular need to have a separate vote for each artifact. Similarly as in case of providers - where we vote on many artifacts and we can verify and test a number of them. And If we find a problem in some of them, release manager can remove some of the artifacts from the release and ask for RC2 for only those fixed artifacts.

Nowhere in the release procedure of the ASF it is written that the vote should be on one artifact at a time.

But in this case I think we have another problem to solve.

I think all our API clients should come directly from airlow  they shoudl all come even from the same source package. The source package of Airflow should be enough to generate not only Apache Airflow package, but also the clients. 

This is another point that I am raising in https://github.com/apache/airflow-client-go/pull/28

Comment (potiuk) [reply dev/README_RELEASE_AIRFLOW.md:45]: And another example of issue that is likely caused by outdated client of ours (thought it is hard to say because it almost looks like no-one actually cares about the Python client so far). 

https://github.com/apache/airflow/pull/28521

I hope by getting this one sorted out will help us to avoid the situation where the client is sort of abandonded even if we as a community release it.

I believe bringing it to the repo, and making it part of the release process is the only way we can improve the situation, otherwise it looks like we are neglecting our users - we have an API that was highly requested, which is really part of the "core" offering of Airflow but we neglect it. 

That situation actually makes me think that we should possibly NOT split providers to separate repositories because we might end up in e similar situation. I will certainly use this as an example when we discuss it. But that's a side comment only.

Comment (pierrejeambrun) [reply dev/README_RELEASE_AIRFLOW.md:45]: I just updated and rebased the PR.

The change aim to first generate rc candidates of the API clients if relevant so we can vote against them. (Add a small section in the dev mail so we can know clients rc are generated).

Publish to PyPI finals version without rc suffix based on the voting artifacts.

This calls for improvement when we will get more familiar with the clients release process, for instance:
- Add the API clients package to the 'notify developers' email. (with the PyPI link)
- Upload artifacts to the `https://dist.apache.org/repos/dist/dev/airflow/`. The folder exists for clients but is empty for python, and Go is missing.

Comment (potiuk) [reply dev/README_RELEASE_AIRFLOW.md:45]: Agree. We should get to follow the reelease process and figure out all the edge case specifics. We were (are) doing it continuously. Pretty much every time we find missing/wrong pieces of the process to follow and we find out what needs to be improved - we've done through that with @eladkal for providers and that was a really helpful exercise to add some really good clarifications and improvements.

BTW. maybe that is a good idea @pierrejeambrun you do the next release for Airflow with those instructions in? Similarly to what we've done with @eladkal on providers that might be helpful to go through that with @ephraimbuddy and others explaining all the questions and doubts, so that we could add some clarifications and get rid of some assumptions we have in our heads. This is encouraged by the ASF release management policies to have rotation for release management duty and people taking turns and the release manager does not have to be a PMC member - being commiter is enough https://www.apache.org/legal/release-policy.html

**Review by potiuk — APPROVED** [review]

File: dev/README_RELEASE_AIRFLOW.md
Comment (pierrejeambrun) [reply dev/README_RELEASE_AIRFLOW.md:45]: I would love to do the next release if possible. (Or at least assist if preferable).

I will wait a day or so before merging to give Elad time to see the last changes.

Comment (potiuk) [reply dev/README_RELEASE_AIRFLOW.md:45]: @ephraimbuddy - since you are usually release manager of Airflow - we have not seen your comments/questions/suggestions here ? What do you think about that (because you will be - likely following it in the next release).

Comment (eladkal) [reply dev/README_RELEASE_AIRFLOW.md:45]: I love the idea of binding clients with Airflow core release.

**Review by eladkal — APPROVED** [review]

File: dev/README_RELEASE_AIRFLOW.md
Comment (ephraimbuddy) [reply dev/README_RELEASE_AIRFLOW.md:45]: Sorry for not commenting. Here's my take:
Releasing the API client together with Airflow sounds good but how do we determine that an API client should be released for a particular Airflow patch release? It's not clear from this PR, I think we should state it clearly. 
Also, I haven't looked at the API client release process, did we document the processes for their release somewhere? 
@pierrejeambrun I will ping you so we both work on the next release when it's time.

Comment (potiuk) [reply dev/README_RELEASE_AIRFLOW.md:45]: The release process (for python client) is here: https://github.com/apache/airflow-client-python#release-process for go client is here: https://github.com/apache/airflow-client-go#release-process

I think the we can determine the need of building it comes from:

a) changes/fixes in openapi specification (internal)
b) available releases of newer version of open-api generator (external) 

If we all agree this is a good idea, I am also happy to move the whole generation as a automated CI process (so that we automatically check out the latest generator and run the package build entirely from `apache-airflow` repository - also for Go Client (as discussed in https://github.com/apache/airflow/pull/28521#discussion_r1059948166 I think the currrent approach is not really good because we commit generated code without being able to check the code).

Comment (ephraimbuddy) [reply dev/README_RELEASE_AIRFLOW.md:45]: Thanks, @potiuk . The idea is good. I fully support it

Comment (pierrejeambrun) [reply dev/README_RELEASE_AIRFLOW.md:45]: Sounds good to me !

I just updated the PR to add the links to the clients release process that Jarek mentioned.

> how do we determine that an API client should be released for a particular Airflow patch release?

Fair point. Just updated that :)

Thanks for the feedback, and looking forward to helping on the next release :)

File: dev/README_RELEASE_AIRFLOW.md
Comment (ephraimbuddy) [inline dev/README_RELEASE_AIRFLOW.md]:
```suggestion
If API clients are to be released in this airflow version:
```

File: dev/README_RELEASE_AIRFLOW.md
Comment (ephraimbuddy) [reply dev/README_RELEASE_AIRFLOW.md:45]: cc @jedcunningham WDYT about this

Comment (jedcunningham) [reply dev/README_RELEASE_AIRFLOW.md:45]: No objections here.

Comment (pierrejeambrun) [reply dev/README_RELEASE_AIRFLOW.md:45]: Thanks @jedcunningham, right in time to try it on 2.5.1 :)

**Pull request merged**
